cpp/harvest/marc_issn_lookup.cc

```diff
@@ -0,0 +1,171 @@
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <fstream>
+#include <iostream>
+#include "CORE.h"
+#include "FileUtil.h"
+#include "MARC.h"
+#include "util.h"
+
+namespace {
+[[noreturn]] void Usage() {
+    ::Usage(
+        "marc_input_articles marc_input_journals marc_output_articles"
+        "\n"
+        "- marc_input_artiles is a file containing all article information taken from CORE. \n"
+        "- marc_input_journals is a file containing journal information. Please use issn_lookup.py to generate this file. \n"
+        "- marc_output_articles is an output file generated by this tool. \n");
+
+    std::exit(EXIT_FAILURE);
+}
+struct SubFieldInfo {
+    std::string i;
+    std::string t;
+    std::string w;
+    std::string x;
+    bool is_found;
+
+    SubFieldInfo(bool b) { this->is_found = b; }
+
+    void Builder(MARC::Record &record) {
+        for (auto &field : record) {
+            if (field.getTag() == "001")
+                this->w = "(DE-627)" + field.getContents();
+
+            if (field.getTag() == "773")
+                this->x = field.getContents();
+
+            if (field.getTag() == "245") {
+                MARC::Subfields subfields(field.getSubfields());
+                std::string subfield_a(subfields.getFirstSubfieldWithCode('a'));
+                std::string subfield_b(subfields.getFirstSubfieldWithCode('b'));
+                if (subfield_a != "" && subfield_b != "")
+                    this->t = subfield_a + " " + subfield_b;
+                else if (subfield_a != "" && subfield_b == "")
+                    this->t = subfield_a;
+                else if (subfield_a == "" && subfield_b != "")
+                    this->t = subfield_b;
+                else
+                    this->t = "";
+            }
+        }
+        this->i = "In:";
+        this->is_found = true;
+    }
+};
+
+bool IsOnlineISSN(const MARC::Record &record) {
+    for (auto &field : record) {
+        if (field.getTag() == "300") {
+            const std::string sub_field_a(field.getFirstSubfieldWithCode('a'));
+            if (sub_field_a == "Online-Ressource")
+                return true;
+            else
+                return false;
+        }
+    }
+    return false;
+}
+
+SubFieldInfo LookupISSNInDataSouce(const std::string input_journal_filename, const std::string issn) {
+    auto data_source(MARC::Reader::Factory(input_journal_filename));
+    SubFieldInfo sub_field_info = SubFieldInfo(false);
+
+    while (MARC::Record record = data_source->read()) {
+        for (auto &field : record) {
+            if (field.getTag() == "022") {
+                const std::string sub_field_a(field.getFirstSubfieldWithCode('a'));
+                if ((sub_field_a != "") && (sub_field_a == issn)) {
```

Comment:
Here as well: `(sub_field_a != "")` => `(not sub_field_a.empty())`. Same a few lines below for `sub_field_l`.